### PR TITLE
sql: allow parsing of CREATE [TEMP] TABLE ... ON COMMIT

### LINKED ISSUE #46555: sql: on commit rows needs an unimplemented error
ON COMMIT DELETE ROWS specifies that the data are removed from the temporary table at the end of each transaction. We do not yet support this feature:

```SQL
CREATE TEMP TABLE temp_cities
  (
     name VARCHAR(80)
  )
  ON COMMIT DELETE ROWS;
invalid syntax: statement ignored: at or near "on": syntax error
SQLSTATE: 42601
DETAIL: source SQL:
CREATE TEMP TABLE temp_cities
  (
     name VARCHAR(80)
  )
  ON COMMIT DELETE ROWS
  ^
```
This should return an unimplemented feature error for on commit rows. 

we also need on on commit drop and preserve rows:
```
CREATE TEMPORARY TABLE temp_isbns (
    title varchar(255) NOT NULL,
    isbn varchar(255) NOT NULL,
    author_name varchar(255) NOT NULL
) ON COMMIT DROP;
```
More info can be found in postgres here: https://www.postgresql.org/docs/9.3/sql-createtable.html

### PR DESCRIPTION
**Pull request by otan**
Resolves https://github.com/cockroachdb/cockroach/issues/46555.

PostgreSQL allows temp tables to have custom behavior ON COMMIT to drop
the table or delete the rows that have been added. We do not support
this behaviour, but we do for the default (PRESERVE ROWS).

Add parsing for ON COMMIT, erroring for the DELETE ROWS and DROP case.
Furthermore, add the same error message as PostgreSQL if a non-temporary
table has the ON COMMIT keyword.

Release justification: low risk, high benefit changes to existing
functionality

Release note (sql change): Allow the ON COMMIT syntax for CREATE TABLE.

### CODE REVIEW

File: docs/generated/sql/bnf/create_table_as_stmt.bnf [hunk 1]
```diff
@@ -1,5 +1,5 @@
 create_table_as_stmt ::=
```
Comment (knz) [inline docs/generated/sql/bnf/create_table_as_stmt.bnf]: can you modify docgen's `diagrams.go` to remove the clause. we're not intending to document it I think.

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -4318,6 +4324,24 @@ opt_table_with:
...
   }
 
+opt_create_table_on_commit:
+  /* EMPTY */
+  {
+    $$.val = tree.CreateTableOnCommitUnset
+  }
+| ON COMMIT PRESERVE ROWS
+  {
+    $$.val = tree.CreateTableOnCommitPreserveRows
```
Comment (knz) [inline pkg/sql/parser/sql.y]: probably `/* SKIP DOC */` here and below

**Review by knz — COMMENTED** [review]
I'll let Rohan comment on correctnes, but I'd like to challenge y'all to start thinking about this question:

*how can we teach current and future SQL team members to systematically return error objects with linked issues* for unimplemented / partially implemented features?

It's not sustainable for me to catch this all the time. We need a better way.

File: pkg/sql/create_table.go [hunk 3]
```diff
@@ -187,6 +187,14 @@ func (n *createTableNode) startExec(params runParams) error {
 		telemetry.Inc(sqltelemetry.CreateTempTableCounter)
 	}
 
+	// TODO(#46556): support ON COMMIT DROP | DELETE ROWS
+	if !isTemporary && n.n.OnCommit != tree.CreateTableOnCommitUnset {
```
Comment (knz) [inline pkg/sql/create_table.go]: `return unimplemented.NewWithIssue(46556, ....)`

Comment (otan) [reply pkg/sql/create_table.go]: This is never intended to be implemented - it is an error as `ON COMMIT` only works on temporary table.

This matches Postgres.

Comment (knz) [reply pkg/sql/create_table.go]: what is the TODO about then?

Comment (otan) [reply pkg/sql/create_table.go]: let me expand on it :)

File: docs/generated/sql/bnf/create_table_as_stmt.bnf (hunk 1 above)
Comment (otan) [reply docs/generated/sql/bnf/create_table_as_stmt.bnf]: i think the SKIP DOC should do it.

File: pkg/sql/parser/sql.y (hunk 2 above)
Comment (otan) [reply pkg/sql/parser/sql.y]: done!

**Comment by otan** [comment]
> how can we teach current and future SQL team members to systematically return error objects with linked issues for unimplemented / partially implemented features?

ML/AI 🤓 

on a serious note, i think this is a just a thing for knowledge sharing and reinforcement. i'm trying to think "can we add a linter for it", and the answer is "probably for sql.y" but not inside the go files.

File: pkg/sql/create_table.go [hunk 4]
```diff
@@ -187,6 +187,18 @@ func (n *createTableNode) startExec(params runParams) error {
...
+		return pgerror.Newf(
+			pgcode.InvalidTableDefinition,
+			"ON COMMIT can only be used on temporary tables",
+		)
+	}
+
+	// TODO(#46556): support ON COMMIT DROP and DELETE ROWS on TEMPORARY TABLE.
+	// If we do this, the n.n.OnCommit variable should probably be stored on the
+	// table descriptor.
+	// Note UNSET / PRESERVE ROWS behave the same way so we do not need to do that for now.
```
Comment (knz) [inline pkg/sql/create_table.go]: Can you assert here that the clause has the default value
(This way if/when alternate possible values are introduced in the parser, we don't start accepting them silently with invalid semantics)

File: pkg/sql/create_table.go [hunk 5]
```diff
@@ -185,6 +185,21 @@ func (n *createTableNode) startExec(params runParams) error {
...
 		telemetry.Inc(sqltelemetry.CreateTempTableCounter)
+
+		// TODO(#46556): support ON COMMIT DROP and DELETE ROWS on TEMPORARY TABLE.
+		// If we do this, the n.n.OnCommit variable should probably be stored on the
+		// table descriptor.
+		// Note UNSET / PRESERVE ROWS behave the same way so we do not need to do that for now.
+		switch n.n.OnCommit {
+		case tree.CreateTableOnCommitUnset, tree.CreateTableOnCommitPreserveRows:
+		default:
+			return errors.AssertionFailedf("ON COMMIT value %d is unrecognized", n.n.OnCommit)
```
Comment (knz) [inline pkg/sql/create_table.go]: isn't one of these cases a candidate for an uimplemented error?

Comment (otan) [reply pkg/sql/create_table.go]: i'm open to either way - the way i see it to get here you must delete the error from the parser, or make the parser allow arbitrary value.

Comment (knz) [reply pkg/sql/create_table.go]: gotcha. I'm fine with this.
(although now I could see us returning an unimplemented error in the parser)

Comment (rohany) [reply pkg/sql/create_table.go]: I'd prefer handling the given cases in the parser and returning unimplemented errors. There are only a few behaviors here, why do we need to handle arbitrary string input?

Comment (otan) [reply pkg/sql/create_table.go]: We aren't handling it here - this is a guard against someone accidentally putting a new enum. And forgetting to do the TODO here.

Comment (otan) [reply pkg/sql/create_table.go]: The Parser will error on all expected cases.

Comment (rohany) [reply pkg/sql/create_table.go]: Sorry, i misunderstood the discussion here.

File: pkg/sql/parser/sql.y [hunk 6]
```diff
@@ -4268,7 +4272,7 @@ create_schema_stmt:
 // WEBDOCS/create-table.html
 // WEBDOCS/create-table-as.html
```
Comment (rohany) [inline pkg/sql/parser/sql.y]: Can you update the help message for create table?

**Comment by rohany** [comment]
LGTM aside from missing help

**Review by knz — APPROVED** [review]
LGTM ditto rohan

> Reviewed 2 of 10 files at r1, 2 of 7 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46594#-:-M3MOboK51qrOhVYv9Pq:b-54hl1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

**Comment by otan** [comment]
bors r=rohany,knz

**Pull request merged**